exchange.go

```diff
@@ -442,7 +475,144 @@ func (re *RabbitExchangeImpl) Receive(exchange ExchangeSettings, queue QueueSett
+			defer closer()
+			batch := make([]amqp.Delivery, 0, queue.Prefetch)
+			timerDuration := time.Duration(maxWait) * time.Millisecond
+			fillWaitTimer := time.NewTimer(timerDuration)
+			handleMessages := func(messages []amqp.Delivery) {
+				if len(messages) == 0 {
+					return
+				}
+				ctx, cancel := context.WithCancel(context.Background())
+				defer cancel()
+				errors := handler(ctx, messages)
+				erroredMessages := make(map[string]struct{})
+				if len(errors) > 0 {
+					for _, msgError := range errors {
+						m := msgError.message
+						err := msgError.err
+						erroredMessages[m.MessageId] = struct{}{}
+						log.Printf("Error handling rabbit message Exchange: %s Queue: %s Body: [%s] %+v\n", exchange.Name, queue.Name, m.Body, err)
+						isProcessingError := IsTemporaryError(err)
+						err = m.Nack(false, false)
+						if err != nil {
+							log.Printf("Error Nack rabbit message %+v\n", err)
+						}
+						if isProcessingError {
+							requeuesObj, ok := m.Headers[requeueHeaderKey]
+							if !ok {
+								continue
+							}
+							noRequeues, ok := requeuesObj.(int32)
+							if ok && noRequeues < maxRequeueNo {
+								noRequeues++
+								err = requeueMessage(
+									channel,
+									queue,
+									exchange.Name,
+									retryExchangeName,
+									m.Body,
+									int(noRequeues),
+								)
+								if err != nil {
+									log.Printf("Error requeueing message %+v\n", err)
+								}
+							}
+						}
+					}
+					for _, m := range messages {
+						if _, ok := erroredMessages[m.MessageId]; !ok {
+							m.Ack(false)
+						}
+					}
+				} else {
+					// If none of the messages resulted in errors, ack multiple (the whole prefetched batch)
+					err = messages[0].Ack(true)
+					if err != nil {
+						log.Printf("Error Ack rabbit message %+v\n", err)
+						return
+					}
+				}
+			}
+			for {
+				// If the `msgs` channel is no longer valid, then we need to open a new one
+				// If that attempt fails, the channel will remain invalid, so we will try again, until we succeed
+				if msgs == nil {
+					channel, msgs, errChan, closer, err = re.getChannel(exchange, queue, retryExchangeName)
+					if err != nil {
+						log.Printf("Cannot get a new channel, after message chanel got closed %+v\n", err)
+						time.Sleep(time.Second)
+					}
+				}
+				// If `msgs` is still empty, we need to retry, so we loop
+				if msgs == nil {
+					continue
+				}
+				select {
+				case m, ok := <-msgs:
+					// if the channel is closed, we want to stop receiving on this one, and we need to open a new one
+					if !ok {
+						msgs = nil
+						continue
+					}
+					batch = append(batch, m)
+					if len(batch) == queue.Prefetch {
+						batchCopy := batch
+						go handleMessages(batchCopy)
+						batch = make([]amqp.Delivery, 0, queue.Prefetch)
+						fillWaitTimer.Reset(timerDuration)
+					}
+				case <-fillWaitTimer.C:
+					if len(batch) > 0 {
+						// Not sure if this is needed
+						batchCopy := batch
```

Comment:
```suggestion
						batchCopy := batch
```

It is needed, the next loop might write into batch (line 585) while we are still processing (line 596).